tests/test_particle_collections.py

```diff
@@ -0,0 +1,55 @@
+from parcels import FieldSet, ScipyParticle, JITParticle
+from parcels import ParticleSetSOA
+import numpy as np
+import pytest
+
+ptype = {'scipy': ScipyParticle, 'jit': JITParticle}
+psettype = {'soa': ParticleSetSOA}
+
+
+def fieldset(xdim=40, ydim=100):
+    U = np.zeros((ydim, xdim), dtype=np.float32)
+    V = np.zeros((ydim, xdim), dtype=np.float32)
+    lon = np.linspace(0, 1, xdim, dtype=np.float32)
+    lat = np.linspace(-60, 60, ydim, dtype=np.float32)
+    depth = np.zeros(1, dtype=np.float32)
+    data = {'U': np.array(U, dtype=np.float32), 'V': np.array(V, dtype=np.float32)}
+    dimensions = {'lat': lat, 'lon': lon, 'depth': depth}
+    return FieldSet.from_data(data, dimensions)
+
+
+@pytest.fixture(name="fieldset")
+def fieldset_fixture(xdim=40, ydim=100):
+    return fieldset(xdim=xdim, ydim=ydim)
+
+
+@pytest.mark.parametrize('pt', ['soa'])
+def test_pset_iteration_forward(fieldset, pt, npart=10):
+    pset = psettype[pt](fieldset, lon=np.linspace(0, 1, npart), lat=np.zeros(npart), pclass=JITParticle)
+    assert np.all(np.diff(np.array([p.id for p in pset])) > 0)
+
+
+@pytest.mark.parametrize('pt', ['soa'])
+def test_pset_iteration_backward(fieldset, pt, npart=10):
+    pset = psettype[pt](fieldset, lon=np.linspace(0, 1, npart), lat=np.zeros(npart), pclass=JITParticle)
+    assert np.all(np.diff(np.array([p.id for p in reversed(pset)])) < 0)
+
+
+@pytest.mark.parametrize('pt', ['soa'])
+def test_pset_get_single_by_index(fieldset, pt, npart=10):
+    pset = psettype[pt](fieldset, lon=np.linspace(0, 1, npart), lat=np.zeros(npart), pclass=JITParticle)
+    assert np.all(np.diff([pset._collection.get_single_by_index(i).lon for i in range(npart)]))
```

Comment:
This comment also applies for the test(s) below.
I think these tests are fine, they test if we get 10 different particles. Which implies we encounter all ten particles at some point when retrieving them by index.
It would be nice to have a test that actually checks if the particles have the value we expect (so using e.g. `np.isclose()`). For the SOA implementation at least we should be able to predict what index / ID will be assigned to a particle, so we could actually do that. However, I am not sure we would also be able to do this for datastructures that do not internally work with indices (linked lists, trees). Then again, we might choose not to implement the get_by_index function for those at all...
For IDs it does generalize, I think. A particle that is added later, will have a higher ID. So sorting the list of IDs should allow for the prediction of the actual values for each particle.

To expand the coverage of the test even further, you could consider calling `pset.collection.get(i)`. This should forward the function call to the appropriate get function based on the type of the argument. So if `i` is a 32-bit integer, it will use `get_single_by_index` in the background (64-bit integer for IDs), and we get a test of the forwarding "for free". But this is just an idea.

In general I would say when referring to the collection, it would be better to use `pset.collection` rather than `pset._collection`. Both work, as a property is defined within the baseparticleset for the collection. But using `pset.collection` would set the right example, just in case somebody looks to the tests for inspiration.